File: top_interview_questions/1_easy/1_array/two_sum/solution.py
class Solution:
    def twoSum(self, numbers, target):
        complements = {}

        for index, number in enumerate(numbers):
            complements[target - number] = index

        for index, number in enumerate(numbers):
            if number in complements and complements[number] != index:
                return [index, complements[number]]

File: top_interview_questions/1_easy/1_array/two_sum/solution.go
package main

import "fmt"

func main() {
	fmt.Println(twoSum([]int{2, 7, 11, 15}, 9)) // expected []int{0, 1}
}

func twoSum(nums []int, target int) []int {
	complements := make(map[int]int)

	for i, num := range nums {
		complements[target-num] = i
	}

	for j, num := range nums {
		i, exists := complements[num]

		if !exists {
			continue
		}

		if i != j {
			return []int{i, j}
		}
	}

	return []int{-1, -1}
}
